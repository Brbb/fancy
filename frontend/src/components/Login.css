.Login {
  color: cornflowerblue;
  vertical-align: middle;
  text-align: center;
  height: 350px;
  width: 350px;
  background: white;
  position: absolute;
  top: 30%;
  left: 35%;
}

.form {
  display: flex;
  flex-direction: column;
  height: 100%;
  align-items: center;
}
