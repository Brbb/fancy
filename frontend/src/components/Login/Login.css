.Login {
  height: 350px;
  width: 350px;
}


.centered {
  position: fixed;
  top: 40%;
  left: 50%;
  transform: translate(-50%, -40%);
}