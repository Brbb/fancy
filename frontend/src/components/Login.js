import React, { Component } from "react";
import { Link } from "react-router-dom";
import "./Login.css";
import auth from "../services/auth/api";

class Login extends Component {
  constructor() {
    super();
    this.state = {
      username: "",
      password: "",
      error: ""
    };
  }

  dismissError = () => {
    this.setState({ error: "" });
  };

  handleUserChange = event => {
    this.setState({ username: event.target.value });
  };
  handlePassChange = event => {
    this.setState({ password: event.target.value });
  };

  login = async () => {
    var loginResult = await auth.login(
      this.state.username,
      this.state.password
    ); // SSL
    if (!loginResult.err) {
      this.props.history.push("/settings");
    } else {
      this.setState({ error: loginResult.err });
    }
  };

  render() {
    return (
      <div className="Login formContainer">
        <form className="form" onSubmit={this.handleSubmit}>
          <label className="formTitle">Fancy App</label>
          <label className="textFieldLabel">Email</label>
          <input
            type="text"
            className="textField"
            placeholder="user@email.com"
            data-test="username"
            value={this.state.username}
            onChange={this.handleUserChange}
          />

          <label className="textFieldLabel">Password</label>
          <input
            type="password"
            placeholder="super secret"
            data-test="password"
            className="textField"
            value={this.state.password}
            onChange={this.handlePassChange}
          />
          {this.state.error && (
            <label
              className="loginError"
              data-test="error"
              onClick={this.dismissError}
            >
              {this.state.error}
            </label>
          )}
          <input
            className="submitBtn"
            type="button"
            value="Log In"
            onClick={this.login}
            disabled={this.state.username === "" || this.state.password === ""}
          />
          <button className="genericBtn">
            <Link to="/new">Create New Account</Link>
          </button>
        </form>
      </div>
    );
  }
}

export default Login;
