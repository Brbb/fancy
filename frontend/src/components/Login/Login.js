import React, { Component } from "react";
import "./Login.css";
import authApi from "../../services/auth/api";
import userApi from "../../services/users/api";
import {
  Button,
  InputField,
  Message,
  SectionTitle
} from "../Elements/Basics";

class Login extends Component {
  constructor() {
    super();
    this.state = {
      username: "",
      password: "",
      error: ""
    };

    //this.loadUser = this.loadUser.bind(this);
    this.login = this.login.bind(this);
  }

  loadUser = async userId => {
    let user = await userApi.getById(userId);
    if (!user.err) {
      this.props.history.push({
        pathname: "/settings",
        state: { user: user }
      });
    } else {
      this.props.history.push("/error");
    }
  };

  dismissError = () => {
    this.setState({ error: "" });
  };

  handleUserChange = event => {
    this.setState({ username: event.target.value });
  };
  handlePassChange = event => {
    this.setState({ password: event.target.value });
  };

  async login() {
    var loginResult = await authApi.login(
      this.state.username,
      this.state.password
    ); // SSL
    if (!loginResult.err) {
      localStorage.setItem("jwt", loginResult.token);
      localStorage.setItem("me", loginResult.userid);
      await this.loadUser(loginResult.userId);
    } else {
      this.setState({ error: loginResult.err });
    }
  }

  render() {
    return (
      <div className="f-section bordered centered">
        <SectionTitle text="Login" />
        <InputField
          placeholder="user@email.com"
          value={this.state.username}
          onChange={this.handleUserChange}
          text="Email"
          name="emailField"
        />
        <InputField
          type="password"
          placeholder="super secret"
          value={this.state.password}
          onChange={this.handlePassChange}
          text="Password"
          name="passwordField"
        />

        <Button
          name="login-button"
          className="f-submit-button"
          text="Login"
          onClick={this.login}
          disabled={this.state.username === "" || this.state.password === ""}
        />

        <Button
          name="create-button"
          className="f-button"
          text="Create new account"
          onClick={() => {
            this.props.history.push("/new");
          }}
        />

        <Message text={this.state.error} success={this.state.error === ""} />
      </div>
    );
  }
}

export default Login;
