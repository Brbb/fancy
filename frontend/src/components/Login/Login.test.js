import Login from "./Login";
import authApi from "../../services/auth/api";
import userApi from "../../services/users/api";

describe("Login component", () => {
  describe("Rendering tests", () => {
    test("Login shallow-renders correctly", () => {
      let wrapper = shallow(<Login />);
      expect(wrapper).toMatchSnapshot();
    });

    test("Login mounts and evaluates state correctly", () => {
      let wrapper = mount(<Login />);
      wrapper
        .find('[name="emailField"]')
        .last()
        .simulate("change", { target: { value: "testemail@provider.com" } });

      expect(wrapper).toMatchSnapshot();
      expect(wrapper.state().username).toEqual("testemail@provider.com");
    });
  });

  describe("Props tests", () => {
    test("Login button is disabled if one InputField is empty", () => {
      let wrapper = shallow(<Login />);
      wrapper.setState({ username: "test" });
      const loginButton = wrapper.find("Button").first();
      expect(loginButton.prop("name")).toEqual("login-button");
      expect(loginButton.prop("disabled")).toBeTruthy();
    });

    test("Login button is enabled if the InputFields have value", () => {
      let wrapper = shallow(<Login />);
      wrapper.setState({ username: "test", password: "1234" });
      const loginButton = wrapper.find("Button").first();
      expect(loginButton.prop("name")).toEqual("login-button");
      expect(loginButton.prop("disabled")).toBeFalsy();
    });
  });

  describe("Navigation test", () => {
    test("Create account button redirects to /new", () => {
      const locationMock = { pathname: location.pathname };

      const historyMock = {
        push: jest.fn(() => {
          locationMock.pathname = "/new";
        })
      };
      let wrapper = shallow(<Login history={historyMock} />);
      const createButton = wrapper.find("[name='create-button']").first();
      createButton.simulate("click");

      expect(locationMock.pathname).toEqual("/new");
      expect(historyMock.push.mock.calls.length).toEqual(1);
    });

    test("Correct login navigates to <Settings />", () => {
      const locationMock = { pathname: location.pathname };

      const historyMock = {
        push: jest.fn(payload => {
          locationMock.pathname = payload.pathname;
          locationMock.state = payload.state;
        })
      };
      let wrapper = mount(<Login history={historyMock} />);
      const mockUser = function(id) {
        return {
          settings: {
            language: "Chinese",
            isPrivate: false
          },
          email: "asd",
          _id: id
        };
      };
      jest.spyOn(userApi, "getById").mockImplementation(id => {
        return Promise.resolve(mockUser(id));
      });

      expect(userApi.getById).toBeCalledTimes(0);
      wrapper
        .instance()
        .loadUser("abcdf123")
        .then(() => {
          expect(userApi.getById).toBeCalledTimes(1);
          expect(locationMock.pathname).toEqual("/settings");
          expect(locationMock.state.user).toEqual(mockUser("abcdf123"));
        });
    });
  });

  describe("Login API call result (MOCK)", () => {
    test("Backend returns error and <Login /> displays it properly", () => {
      let wrapper = mount(<Login />);

      jest.spyOn(authApi, "login").mockImplementation(() => {
        return Promise.resolve({ err: "anyerror" });
      });

      wrapper.setState({ username: "test", password: "12345" });
      const loginButton = wrapper.find("[name='login-button']").first();

      return loginButton
        .prop("onClick")()
        .then(() => {
          expect(wrapper.state().error).toEqual("anyerror"); //checking if the state has been updated correctly
          wrapper.update();
          const messageElement = wrapper.find("Message").first(); //checking if the Message component will show the error
          expect(messageElement.prop("text")).toEqual("anyerror");
          expect(messageElement.prop("success")).toBeFalsy();
          jest.clearAllMocks();
        });
    });

    test("Backend returns correct authentication and <Login /> calls loadUser", () => {
      jest.spyOn(authApi, "login").mockImplementation(() => {
        return Promise.resolve({ token: "asd123", userid: 12345 });
      });

      let wrapper = mount(<Login />);
      const spyLoadUser = jest
        .spyOn(wrapper.instance(), "loadUser")
        .mockImplementationOnce(() => {
          /*empty*/
        });

      wrapper.setState({ username: "test", password: "12345" });
      const loginButton = wrapper.find("[name='login-button']").first();

      expect(spyLoadUser).toHaveBeenCalledTimes(0);

      return loginButton
        .prop("onClick")()
        .then(() => {
          expect(spyLoadUser).toHaveBeenCalledTimes(1);
          jest.clearAllMocks();
        });
    });
  });
});
