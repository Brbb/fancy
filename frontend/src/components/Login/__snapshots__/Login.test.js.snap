// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login component Rendering tests Login mounts and evaluates state correctly 1`] = `
<div
  className="f-section bordered centered"
>
  <div
    className="f-section-title"
  >
    <label>
      Login
    </label>
  </div>
  <div
    className="f-input"
  >
    <label
      className="f-input-field-label"
    >
      Email
    </label>
    <input
      className="f-input-field"
      name="emailField"
      onChange={[Function]}
      placeholder="user@email.com"
      value="testemail@provider.com"
    />
  </div>
  <div
    className="f-input"
  >
    <label
      className="f-input-field-label"
    >
      Password
    </label>
    <input
      className="f-input-field"
      name="passwordField"
      onChange={[Function]}
      placeholder="super secret"
      type="password"
      value=""
    />
  </div>
  <input
    className="f-submit-button"
    disabled={true}
    name="login-button"
    onClick={[Function]}
    type="button"
    value="Login"
  />
  <input
    className="f-button"
    name="create-button"
    onClick={[Function]}
    type="button"
    value="Create new account"
  />
  <label
    className="f-message f-message-s"
  />
</div>
`;

exports[`Login component Rendering tests Login shallow-renders correctly 1`] = `
<div
  className="f-section bordered centered"
>
  <SectionTitle
    text="Login"
  />
  <InputField
    name="emailField"
    onChange={[Function]}
    placeholder="user@email.com"
    text="Email"
    value=""
  />
  <InputField
    name="passwordField"
    onChange={[Function]}
    placeholder="super secret"
    text="Password"
    type="password"
    value=""
  />
  <Button
    className="f-submit-button"
    disabled={true}
    name="login-button"
    onClick={[Function]}
    text="Login"
  />
  <Button
    className="f-button"
    name="create-button"
    onClick={[Function]}
    text="Create new account"
  />
  <Message
    success={true}
    text=""
  />
</div>
`;
